### ISSUE TITLE
Add support to the HiGHS solver MathOpt

### ISSUE THREAD

**Issue by SanPen** [google/or-tools#3116]
Hi, 

[HiGHS](https://github.com/ERGO-Code/HiGHS) is a magnificent open-source solver that's able to run in parallel, clearly beating CBC in some of the tests I've done in python.

Is there any plan to integrate it into or-tools? (C++)

If not, I'd like to do it myself since the interface looks similar to CBC's. What would be the right way to start?

thanks,
Santiago

Mizux's EDIT(to have task list need to edit the first message...):

* [x] Create github issue label
* [x] Incorporate it to CMake build (see below)
  * [x] Build HiGHS using a CMake based build (provided by project, need to test)
  * [x] BUILD_HIGHS using cmake `fetch()`
  * [ ] `find_package()` module
    * [ ] Test system wide install (https://repology.org/project/highs/badges) 
* [x] Incorporate it to bazel build
  * [x] Build HIGHS using bazel
* [x] Integrate to MathOpt

note: my dev will be in `[mizux/highs]`  (to create) branch and regularly rebased on top of `[master]`

**Comment by lperron** [comment]
Yes, but we do not have the time to write the interface for it. So we will wait from some contribution for it.

**Comment by lperron** [comment]
You can copy the cbc_inteface.cc file

**Comment by Mizux** [comment]
Todo list:

* [ ] Add option in CMake to enable support (`OFF` by default maybe `ON` later)
e.g. 
https://github.com/google/or-tools/blob/2cb85b4eead4c38e1c54b48044f92087cf165bce/CMakeLists.txt#L146-L153
* [ ] Add FetchContent() section tot build it
e.g.
https://github.com/google/or-tools/blob/2cb85b4eead4c38e1c54b48044f92087cf165bce/cmake/dependencies/CMakeLists.txt#L155-L174
ref: https://github.com/ERGO-Code/HiGHS/blob/master/CMakeLists.txt
note: need to patch their CMake (no CMake config and install rule, no CTest use etc...)
* [ ] Add A `FindHIGHS.cmake` to find a system-wide installed version
e.g. https://github.com/google/or-tools/blob/stable/cmake/FindGLPK.cmake
* [ ] Add section in `deps.cmake`
e.g. https://github.com/google/or-tools/blob/2cb85b4eead4c38e1c54b48044f92087cf165bce/cmake/deps.cmake#L61-L65
* [ ] Add a define `USE_HIGHS` in `cpp.cmake`
e.g. https://github.com/google/or-tools/blob/2cb85b4eead4c38e1c54b48044f92087cf165bce/cmake/cpp.cmake#L26-L28
* [ ] Link MPSovler build to it
e.g. https://github.com/google/or-tools/blob/2cb85b4eead4c38e1c54b48044f92087cf165bce/ortools/linear_solver/CMakeLists.txt#L26

**Comment by gregy4** [comment]
Any news about support of highs solver ? Do you work on it, SanPen ?
Latest version 1.2.2 solved one of my unsolvable mip problems by any open source solver (I maybe tried them all) including scip. Since optimal solution was found under 3 seconds, I was very satisfied and maybe there’s time to think about usage of mip solver in some cases for me.

**Comment by SanPen** [comment]
My process was the following:
- I went through ortools code and discovered that ortools modelling in C++was not as simple as ortools in python (a + b <= 3, etc) which disappointed me.
- I thought that making my own ortools exactly the way I wanted was less hassle, so I did it. It took me about a week.
- I integrated Highs rather easily.

So, essentially I thought that fighting a code that I don't understand would be far more complicated than building everything from scratch, so I made my own solver interface with matrix-variable operations embedded.

If we find time to do a gathering with the authors and get dummy-proof guide, I might write the Highs interface, otherwise it is too much for me.

**Comment by jajhall** [comment]
Having first got in touch with @lperron in August of last year, the HiGHS team is now in a position to contribute to writing an OR-Tools interface to HiGHS. Not only will this get around the "not open-source to industry" issue of SCIP, but it will give OR-Tools a better MIP solver than SCIP - see http://plato.asu.edu/ftp/milp.html

That said, I'm interested to read about the experiences of @SanPen and @Mizux on this page, and it would be great if you were able to feed your knowledge into the process.

**Comment by jajhall** [comment]
I've flagged up this exchange with @galabovaa and she will have a look at what's involved to implement the proto solver. 

With your help, I'm sure that we'll make this happen

**Comment by Mizux** [comment]
dev note:
To Fix/Patch
* [ ] `highs_interface.cc` (currently only a copy/paste from `scip_interface.cc`)
* [ ] HiGHS CMake fix:
  * [ ] Use of include(CTEST)
  * [ ] Why using `SHARED_BUILD` option then set the CMake orignal one BUILD_SHARED_LIB ...
  * [ ] fast build/regular build code refactoring (`add_subbdirectory(src)` x2 !!!)
  * [ ] Add install rules for config files and Target.cmake

**Comment by jajhall** [comment]
Thanks.  @galabovaa got Bazel to build HiGHS before, so can contribute. She's busy for the next couple of weeks, though

**Comment by frederikgeth** [comment]
This is a post to say hi and to confirm interest in this feature! We currently use CLP and GLOP through Java OR tools. We are grateful to everyone for all the existing contributions and excited about the opportunity to start using HiGHS soon. Let me know if there is an opportunity to help in testing.

**Comment by jajhall** [comment]
@galabovaa is now in a position to start work on an OR-Tools <-> HiGHS interface. Can't say how long it will take, but it's definitely happening!

**Comment by igrqb** [comment]
Awesome! Kudos for the great work :)

**Comment by wxinix** [comment]
any updates about ORTools + Highs?

**Comment by FernandoMotkraft** [comment]
@Mizux Hello, I wonder if you folks are still working on this. I would like to move away from SCIP, due to speed on MIP, also due to potential memory leaks on SCIP. Do you folks still have plans to enable HiGHs within or-tools? Is there an approximate timing you could share?

**Comment by lperron** [comment]
We are, low priority though.
Not in 9.8 (coming soon). Hopefully 9.9
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le mar. 24 oct. 2023 à 16:16, FernandoMotkraft ***@***.***> a
écrit :

**Comment by FernandoMotkraft** [comment]
Thanks for the info @lperron

**Comment by Mizux** [comment]
It seems we already have Highs support for mathopt internally (i.e. blaze only), I need to export and clean it asap.
But as Laurent say we want to release the incoming v9.8 this week first so we won't integrate it before v9.9 (ETA around Christmas)

**Comment by FernandoMotkraft** [comment]
That would be a nice Xmas present @Mizux :)

**Comment by individ2016** [comment]
Hello! Any news about HiGHS integration? :)

**Comment by Mizux** [comment]
Currently cleaning HiGHs build system, the objective is to have HiGHS support working for 9.10 this week or next (finger crossed):
* https://github.com/ERGO-Code/HiGHS/pull/1701

**Comment by Mizux** [comment]
Currently some math_opt tests fail when using HiGHs 1.7.0 (internally we still use 1.4.1) trying to understand what happen

**Comment by jajhall** [comment]
> Currently some math_opt tests fail when using HiGHs 1.7.0 (internally we still use 1.4.1) trying to understand what happen

Happy to help you understand why, if you'll share HiGHS output

**Comment by Mizux** [comment]
> > Currently some math_opt tests fail when using HiGHs 1.7.0 (internally we still use 1.4.1) trying to understand what happen
> 
> Happy to help you understand why, if you'll share HiGHS output

Thanks, I've create the issue #4203 with all information I have, I'm trying to investigate...
